pkg/config/virtio.go

```diff
@@ -181,6 +220,122 @@ func (dev *VirtioSerial) FromOptions(options []option) error {
 	return dev.validate()
 }
 
+// VirtioInputNew creates a new input device for the virtual machine.
+// The inputType parameter is the type of virtio-input device that will be added
+// to the machine.
+func VirtioInputNew(inputType string) (VirtioDevice, error) {
+	dev := &VirtioInput{
+		InputType: inputType,
+	}
+	if err := dev.validate(); err != nil {
+		return nil, err
+	}
+
+	return dev, nil
+}
+
+func (dev *VirtioInput) validate() error {
+	if dev.InputType != VirtioInputPointingDevice && dev.InputType != VirtioInputKeyboardDevice {
+		return fmt.Errorf("Unknown option for virtio-input devices: %s", dev.InputType)
+	}
+
+	return nil
+}
+
+func (dev *VirtioInput) ToCmdLine() ([]string, error) {
+	if err := dev.validate(); err != nil {
+		return nil, err
+	}
+
+	return []string{"--device", fmt.Sprintf("virtio-input,%s", dev.InputType)}, nil
+}
+
+func (dev *VirtioInput) FromOptions(options []option) error {
+	for _, option := range options {
+		switch option.key {
+		case VirtioInputPointingDevice, VirtioInputKeyboardDevice:
+			if option.value != "" {
+				return fmt.Errorf(fmt.Sprintf("Unexpected value for virtio-input %s option: %s", option.key, option.value))
+			}
+			dev.InputType = option.key
+		default:
+			return fmt.Errorf("Unknown option for virtio-input devices: %s", option.key)
+		}
+	}
+	return dev.validate()
+}
+
+// VirtioGPUNew creates a new gpu device for the virtual machine.
+// The usesGUI parameter determines whether a graphical application window will
+// be displayed
+func VirtioGPUNew() (VirtioDevice, error) {
+	virtioGPU := &VirtioGPU{
+		Device:  VirtioGPUDisplayDevice,
+		UsesGUI: false,
+	}
+	virtioGPU.VirtioGPUResolution = VirtioGPUResolution{
+		Height: defaultVirtioGPUResolutionHeight,
+		Width:  defaultVirtioGPUResolutionWidth,
+	}
+	return virtioGPU, nil
+}
+
+func (dev *VirtioGPU) validate() error {
+	if dev.Device != VirtioGPUDisplayDevice {
+		return fmt.Errorf("Unknown option for virtio-gpu devices: %s", dev.Device)
+	}
+
+	if dev.Height < 1 || dev.Width < 1 {
+		return fmt.Errorf("Invalid dimensions for virtio-gpu device resolution: %dx%d", dev.Height, dev.Width)
+	}
+
+	return nil
+}
+
+func (dev *VirtioGPU) ToCmdLine() ([]string, error) {
+	if err := dev.validate(); err != nil {
+		return nil, err
+	}
+
+	return []string{"--device", fmt.Sprintf("virtio-gpu,%s,height=%d,width=%d", dev.Device, dev.Height, dev.Width)}, nil
+}
+
+func (dev *VirtioGPU) FromOptions(options []option) error {
+	for _, option := range options {
+		if option.key != VirtioGPUDisplayDevice && option.key != VirtioGPUResolutionWidth && option.key != VirtioGPUResolutionHeight {
+			return fmt.Errorf("Unknown option for virtio-gpu devices: %s", option.key)
+		}
```

Comment:
This can be the `default:` case in the switch/case below.

Reply:
Fixed.